src/wallet/wallet.cpp

```diff
@@ -3797,7 +3797,187 @@ bool CWallet::CreateCoinStakeFromMine(interfaces::Chain::Lock& locked_chain, con
+
+                if (!keystore.GetKey(delegation.staker, key))
+                {
+                    LogPrint(BCLog::COINSTAKE, "CreateCoinStake : failed to get staker key for kernel type=%d\n", whichType);
+                    break;  // unable to find corresponding public key
+                }
+                scriptPubKeyStaker << key.GetPubKey().getvch() << OP_CHECKSIG;
+                scriptPubKeyHashKernel = scriptPubKeyKernel;
+            }
+            if (whichType == TX_PUBKEY)
+            {
+                valtype& vchPubKey = vSolutions[0];
+                uint160 hash160(Hash160(vchPubKey));
+                CKeyID pubKeyHash(hash160);
+
+                if(!GetDelegation(pubKeyHash, delegation))
+                    return error("CreateCoinStake: Failed to find delegation");
+
+                if (!keystore.GetKey(delegation.staker, key))
+                {
+                    LogPrint(BCLog::COINSTAKE, "CreateCoinStake : failed to get staker key for kernel type=%d\n", whichType);
+                    break;  // unable to find corresponding public key
+                }
+
+                scriptPubKeyStaker << key.GetPubKey().getvch() << OP_CHECKSIG;
+                scriptPubKeyHashKernel = CScript() << OP_DUP << OP_HASH160 << ToByteVector(hash160) << OP_EQUALVERIFY << OP_CHECKSIG;
+            }
+
+            txNew.vin.push_back(CTxIn(prevoutStake));
+            nCredit += coinPrev.out.nValue;
+            nValue = coinPrev.out.nValue;
+            txNew.vout.push_back(CTxOut(0, scriptPubKeyStaker));
+            txNew.vout.push_back(CTxOut(0, scriptPubKeyHashKernel));
+
+            LogPrint(BCLog::COINSTAKE, "CreateCoinStake : added kernel type=%d\n", whichType);
+            fKernelFound = true;
+            break;
+        }
+
+        if (fKernelFound)
+            break; // if kernel is found stop searching
+    }
+
+    if (nCredit == 0)
+        return false;
+
+    const Consensus::Params& consensusParams = Params().GetConsensus();
+    int64_t nRewardPiece = 0;
+    int64_t nRewardStaker = 0;
+    // Calculate reward
+    {
+        int64_t nTotalReward = nTotalFees + GetBlockSubsidy(pindexPrev->nHeight + 1, consensusParams);
+        if (nTotalReward < 0)
+            return false;
+
+        if(pindexPrev->nHeight < consensusParams.nFirstMPoSBlock)
+        {
+            // Keep whole reward
+            nRewardStaker = nTotalReward * 100 / delegation.fee;
+            nCredit += nTotalReward - nRewardStaker;
+        }
+        else
+        {
+            // Split the reward when mpos is used
+            nRewardPiece = nTotalReward / consensusParams.nMPoSRewardRecipients;
+            int64_t nRewardDelegate = nRewardPiece + nTotalReward % consensusParams.nMPoSRewardRecipients;
+            nRewardStaker = nRewardDelegate * 100 / delegation.fee;
+            nCredit += nRewardDelegate - nRewardStaker;
+        }
+    }
+
+    // Set output amount
+    txNew.vout[1].nValue = nRewardStaker;
```

Comment:
Should prevent generating zero value utxo when `delegation.fee=0`

Reply:
0 and 100 fees are allowed in the delegation contract, so if one of those fees are used then one output will be 0 and the other output will contain the full reward. @qtum-neil Does modification is required?